c_glib/arrow-dataset-glib/filter.h

```diff
@@ -0,0 +1,387 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+#pragma once
+
+G_BEGIN_DECLS
+
+struct GADInExpression;
+struct GADCastExpression;
+
+/* arrow::dataset::ExpressionType */
+
+/**
+ * GADExpressionType
+ * @GAD_EXPRESSION_TYPE_FIELD:
+ *   A reference to a column within a record batch,
+ *   will evaluate to an array
+ * @GAD_EXPRESSION_TYPE_SCALAR:
+ *   A literal singular value encapsulated in a Scalar
+ * @GAD_EXPRESSION_TYPE_NOT:
+ *   An inversion of another expression
+ * @GAD_EXPRESSION_TYPE_CAST:
+ *   Cast an expression to a given DataType
+ * @GAD_EXPRESSION_TYPE_AND:
+ *   A conjunction of multiple expressions (true if all operands are true)
+ * @GAD_EXPRESSION_TYPE_OR:
+ *   A disjunction of multiple expressions (true if any operand is true)
+ * @GAD_EXPRESSION_TYPE_COMPARISON:
+ *   A comparison of two other expressions
+ * @GAD_EXPRESSION_TYPE_IS_VALID:
+ *   Extract validity as a boolean expression
+ * @GAD_EXPRESSION_TYPE_IN:
+ *   Check each element for membership in a set
+ * @GAD_EXPRESSION_TYPE_CUSTOM:
+ *   Custom user defined expression
+ */
+typedef enum {
+  GAD_EXPRESSION_TYPE_FIELD,
+  GAD_EXPRESSION_TYPE_SCALAR,
+  GAD_EXPRESSION_TYPE_NOT,
+  GAD_EXPRESSION_TYPE_CAST,
+  GAD_EXPRESSION_TYPE_AND,
+  GAD_EXPRESSION_TYPE_OR,
+  GAD_EXPRESSION_TYPE_COMPARISON,
+  GAD_EXPRESSION_TYPE_IS_VALID,
+  GAD_EXPRESSION_TYPE_IN,
+  GAD_EXPRESSION_TYPE_CUSTOM
+} GADExpressionType;
+
+/* arrow::dataset::Expression */
+
+#define GAD_TYPE_EXPRESSION (gad_expression_get_type())
+G_DECLARE_DERIVABLE_TYPE(GADExpression,
+                         gad_expression,
+                         GAD,
+                         EXPRESSION,
+                         GObject)
+struct _GADExpressionClass
+{
+  GObjectClass parent_class;
+};
+
+gboolean
+gad_expression_equal(GADExpression *expression,
+                     GADExpression *other_expression);
+
+/* TODO: GArrowScalar */
```

Comment:
@kou Should we add `GArrowScalar` for `arrow::Scalar`?

Reply:
IMHO this isn't necessary for the moment as dataset currently only supports filtering on primitive columns

Reply:
Thanks!

It seems that  we don't need to bind `arrow::Scalar` by providing `SchalarExpression` subclases for each scalar type.